src/goto-programs/goto_convert.cpp

```diff
@@ -868,13 +868,11 @@ void goto_convertt::convert_loop_invariant(
   if(invariant.is_nil())
     return;
 
-  goto_programt no_sideeffects;
-  clean_expr(invariant, no_sideeffects, mode);
-
-  INVARIANT_WITH_DIAGNOSTICS(
-    no_sideeffects.instructions.empty(),
-    "loop invariant is not side-effect free",
-    code.find_source_location());
+  if(has_subexpr(invariant, ID_side_effect))
```

Comment:
As I recall, this was the really contentious part last time around.  This seems much less disruptive.

I think throwing the exception is correct because the user can trigger this even if the tool is non-buggy.  I wonder if it would be more user-friendly to check this earlier but ... I guess we need to check it here as well.

I wonder if this might still be a bit too conservative but, if it works for you then fine.

Reply:
Thank you, yes earlier I had removed this part entirely and was doing some checking in `goto-instrument` instead. Now this check is performed at `goto-cc` compilation.

> I wonder if this might still be a bit too conservative but, if it works for you then fine.

I completely agree with you. And it's on our roadmap to allow function calls within contracts, if those functions have an empty assigns clause (so no side effects).